# Lists, Iteration and Loops

## Slicing Sections out of Strings

Characters in a string can be identified from their position using their <span>*index*</span>.  


* Numbering in Python starts from zero. 

* This can be thought of as numbering the positions <span>*between*</span> the letters:
```
| W | o | r | d | ! |
0   1   2   3   4   5
```


We can obtain more than one letter by <span>*slicing*</span> it using a
range of indices: `string[index1:index2]`, where the slice is all
letters <span>*between*</span> the two indices (see diagram above).

In [1]:
a = "Word!"

print(a[0:4])

Word


* **empty indices**:  If we leave either of the two positions (`index1` or `index2`) blank:  
`[:]` means "from the start to the end"  
`[1:]` means "from position 1 to the end"  
`[:3]` means "from the start to position 3" 

* **Use index *slicing* to cut out the sections of the string indicated in the comments:** 

In [2]:
a = "Word!"

# W
print(a[?:?])

SyntaxError: invalid syntax (<ipython-input-2-c3c90b27fd36>, line 4)

In [3]:
# Wor
print(a[?])

SyntaxError: invalid syntax (<ipython-input-3-0933c0c7ea4e>, line 2)

In [ ]:
# d!
print(?)

Now study the example below:

In [ ]:
print(a[:3]+a[3:])

### Accessing a character in a string using its *Index* 

Accessing a single character in a string can be simplified using just first the index (i.e. *before* the letter).  

This is done using the string name with the index number in square brackets `string[index]`.  

In [4]:
a = "Word!"
print(a[0])

W


* **Now do the same using the indices `1,2,3,4,5` in turn**:
    - re-run the same cell each time with a different index:

In [ ]:
print(a[?])

* Note that we cannot access *indices* beyond the length of the character string held by the variable `a`, trying this produces an error.

### Negative Indexing

Negative numbers can be used to start from the end of the string and work backwards:
```
 | B | a | c | k | w | a | r | d | s |
-9  -8  -7  -6  -5  -4  -3  -2  -1
```

In [ ]:
b = "Backwards"
print(b[-5:-1])

In [ ]:
# Slice the word Back using negative indexing
print(b[:-5])

In [ ]:
# Slice the word wards using negative indexing
print(b[-5:])

In [ ]:
# make note of this exampple and later think of how to automate it using loops...
print(b[-1] + b[-2] + b[-3] + b[-4] + b[-5] + b[-6] + b[-7] + b[-8] + b[-9])

## Lists 


A `list` is a data type consisting of multiple elements. 

Lists are defined by enclosing them in square brackets.  
The elements in a list can be of any type in Python,
including text strings or numbers.

In [ ]:
v = [1, 2, 1.5, "Hello"]
print(v)

### Accessing an item in a list using its *Index*

The elements from a list can also be selected using their *index* in
square brackets following the list's name, just like with strings:

In [ ]:
import math

v1 = [1, 2.5, "Hello", "goodbye", math.pi]

In [ ]:
# access the first element
print(v1[0])

In [ ]:
# slice from between the second and fourth "fence-post" (third and fourth elements)
print(v1[2:4])

In [ ]:
# print the last element
print(v1[-1])

### Slicing Lists

We can also *slice* lists to take more than one value at a time.

In [ ]:
pilist = [3, ".", 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 9]

print(pilist[0:7])

### Joining Lists

Concatenating (joining) lists can be done using the `+` symbol, which
does not add the lists in a numerical sense. 

In [ ]:
a = [1, 2, 3, 4, 5]
b = [6, 7, 8, 9, 10]
c = a+b

print(c)

### Nested Lists

Lists can also contain lists. These are known as *nested lists* and are
demonstrated below:

In [1]:
lst1 = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]
lst2 = [[1, 2, 3], lst1, "Hello"]

print(lst2)

[[1, 2, 3], ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'], 'Hello']


In [2]:
ls1 = [1, 2, 3]
ls2 = [4, 5, 6]
ls3 = [7, 8, 9]
lst = [ls1, ls2, ls3]

print(lst)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


### Changing Values in a List 

Any entry in a list can be changed using its index or slice:

In [ ]:
lst1 = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]

# replacing a single element print lst1
lst1[0] = "A"
print(lst1)

In [ ]:
# replacing slice ["d","e","f","g"] with [3,4,5]
lst1[3:7] = [3, 4, 5]
print(lst1)

* replace element "b" with the list [7,8,9]

In [ ]:
lst1[?] = ?
print(lst1)

Expected Result: `['A', [7, 8, 9], 'c', 3, 4, 5, 'h', 'i', 'j']`

Notice the different behaviour when replacing a slice or a single
element with a list:

-   a slice is replaced with elements from the inserted list in place of
    the original slice

-   a single entry has the list inserted *with the list* as the
    *element* entry.

Notice also that the length of the list changes as we replace a 4
element slice with a three slice. Also when counting list entries, the
entries in sub-lists do not get counted, just the sub-list itself as a
whole object.

### Accessing Sub-items of Lists

Look at the following example and see if you can figure out what's going
on (explanation beneath):

In [3]:
lst1 = [1, 1.5, "hello", [1, 2, 3, 4]]

a = lst1[3]
print(a)
print(a[1])

print(lst1[2][1])
print(lst1[3][2:4])

[1, 2, 3, 4]
2
e
[3, 4]


The basic idea is that the first set of square brackets says
which part of the main list we want.  

If this item itself has more than one element we can access these elements  
using another set of square
brackets:  `listname[item][subitem][subsubitem]` etc. 

The first example
above shows this more clearly by firstly extracting item `lst1[3]` of
the main list and then taking element 1 from this. The next example
shows this on a numerical list.

In [1]:
lst = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

print(lst[1])
print(lst[1][0])

a = lst[1][2]
b = lst[2][0]

print(a*b)

[4, 5, 6]
4
42


### Summary

In [5]:
# a comment is ignored

something = "A string of characters"  # giving a text string a variable name
# slicing: remember it goes up to one less than the last index
print(something[2:8])

# accessing using an index number: remember indexing starts at zero
print(something[3])

a_list = ["anything in", 2, "square", "brackets", [1, 2, 3]]

print(a_list[1] * a_list[4][2])  # accessing lists and sub-lists

string
t
6


## Generating New Lists

### Using inbuilt functions

The function `list()` makes a list out of the elements of an object, and `range()` creates a range of numbers for it to list:

In [13]:
# create a list with 10 elements
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [7]:
# Numbers don't have to start at zero:
list(range(5,10))

[5, 6, 7, 8, 9]

In [8]:
# a third parameter controls the step size (integers only)
list(range(1,10,2))

[1, 3, 5, 7, 9]

* Make a list from 3 to 9 in steps of 3 

In [10]:
a = list(range(3,10,3))

print(a)

[3, 6, 9]


Result: `[3, 6, 9]`

### Using *"List Comprehension"*

New lists can be generated by manipulating each of the items in an existing list.  

* For example "print a list of $x^2$ values for all $x$ values **in** the list `[1, 2, 3]`".  

In [11]:
x1 = [1, 2, 3, 4, 5]

x2 = [x**2 for x in x1]
print(x2)

[1, 4, 9, 16, 25]


This method is known as *"list comprehension"*.

This also works for non-numerical lists:

In [12]:
# apply the function len(a) for every word `a` in the list [...]
word_lengths = [len(a) for a in ["spam", "egg", "cheese"]]

print("The words in the list are:", word_lengths, "letters long.")

The words in the list are: [4, 3, 6] letters long.


And for strings:

In [13]:
mystring = "Hello World!"

# reeate a new list with elemets that are the string "The next..." plus each character c in the given list
newlist = ["The next character is "+c for c in mystring]

print(newlist)

['The next character is H', 'The next character is e', 'The next character is l', 'The next character is l', 'The next character is o', 'The next character is  ', 'The next character is W', 'The next character is o', 'The next character is r', 'The next character is l', 'The next character is d', 'The next character is !']


### Exercise: 

Use ***list comprehension*** to create a list with 9 values from $-\pi$ to $\pi$.  
The formula to use is: $-\pi + i(2\pi)/8$ for each value $i$ in a list [0..8] created using `list()` and `range()`. 


* You will need to import `pi` from the `math` library
* The values obtained should be $-\pi, -3\pi/4, -\pi/2, -\pi/4, 0.0, \pi/4, \pi/2, 3\pi/4, \pi$

In [26]:
mylist = (list(range(0,9)))
print (mylist)

from math import pi
pilist = [-pi + i*(2*pi)/8 for i in mylist]
print(pilist)




[0, 1, 2, 3, 4, 5, 6, 7, 8]
[-3.141592653589793, -2.356194490192345, -1.5707963267948966, -0.7853981633974483, 0.0, 0.7853981633974483, 1.5707963267948966, 2.356194490192345, 3.141592653589793]


Hint (double click here to reveal): $\bullet\bullet\bullet$

<!--HINT:
   ```
   OLDLIST = list(range(X)) # creates a list from zero to X-1  
   newlist = [CALCULATION for VALUE in OLDLIST] # suing the formula given in the question
   ```
-->

In [2]:
from math import pi
# Check your answers against this:
what_you_should_get = [-pi, -3*pi/4, -pi/2, -pi/4, 0.0, pi/4, pi/2, 3*pi/4, pi]
print(what_you_should_get)

[-3.141592653589793, -2.356194490192345, -1.5707963267948966, -0.7853981633974483, 0.0, 0.7853981633974483, 1.5707963267948966, 2.356194490192345, 3.141592653589793]


Model Solution (click here to see):

<!--#code below:
from math import pi

nvals = list(range(9))
print(nvals)

pilist = [-pi + i*2*pi/8 for i in ivals]
print(mylist)
-->

* This example shows why counting from zero can be useful (think about this).

## Iterating using FOR loops

Another way to repeatedly persorm some funtion **for** all the items in a list is using a *"`for` loop"*.

### Iterating through Strings 

We can step through the letters in a string and repeat a block of
**INDENTED** code <span>**for**</span> each character in the string. 

Here the `for` statement takes each letter in the string
`"Hello World!"` in turn and assigns it to the variable we call `letr`. 

The indented code block is repeated in a loop for each character. 

Each time the loop repeats the variable `letr` takes the next value.  
Once the string is finished the indented block exits.

* Notice that the code that is repeated has to be *indented* (usually by 4 spaces).

In [ ]:
for letr in "Hello World!":
    print("The next character is: "+letr)

This is equivalent to performing the following commands in sequence:
```python
letr='H'
print("the next letter is:", letr)
letr='e'
print("the next letter is:", letr)
```
... and so on.

### FOR Loops on Lists

We can step through the items in a list in the same way as stepping
through the characters in a string.

In [14]:
for next_item in ["spam", "eggs", "cheese"]:
    print("buy some", next_item)

buy some spam
buy some eggs
buy some cheese


### Performing calculations using FOR loops

Look at the three ways of performing the same task below to see
how FOR loops can be used to simplify a task (the comma stops printing on a new line):

In [15]:
print(1**2, end=', ')
print(2**2, end=', ')
print(3**2)

1, 4, 9


* The `end=', '` argument makes sure each `print()` is followed by a comma and not a new line (return character).

We can rewrite this using a counter `n`:

In [ ]:
n = 1
print(n**2, end=', ')
n = 2
print(n**2, end=', ')
n = 3
print(n**2)

This can now be made compact using a loop:

In [18]:
numbers = [1, 2, 3]
for n in numbers:
    print(n**2, end=', ')

1, 4, 9, 

#### Using the range() Function

The `range` function can be used directly to iterate a block of instructions for
a set number of steps, i.e. for a certain *range* of values.

Here `i` is just a variable name for a counter that takes values in the range `0:5` (i.e. 6 values),  
it could be used in the calculation, or as an index to access a value in a list or string, or not used at all:

* Notice that a **BLOCK** of code is indented by the same number of spaces, and all of this repeats

In [20]:
n = 1

# for each iteration of the loop, times the previous n by 10
for i in range(6):
    n = n*10
    print(i, n, end=', ')
    
print("""\nAnything after the indented block, using normal indentation (back to the margin), 
only gets executed AFTER the loop has finished iterating.""")

0 10, 1 100, 2 1000, 3 10000, 4 100000, 5 1000000, 
Anything after the indented block, using normal indentation (back to the margin), 
only gets executed AFTER the loop has finished iterating.


The syntax is `range(int1,int2,step)`, where the list generated is all
the integers <span>*between*</span> the integers `int1` and `int2` in
steps of `step`.  
This does not include the actual value `int2` (see
slicing rules in Lesson 2).  
If the start value (`int1`)
is omitted the list starts from 0, and the default step is 1.

### Exercise 2

Use a `for` loop to create a list with 9 values from $-\pi$ to $\pi$.  
Use the same formula as before: $-\pi + i(2\pi)/8$. 

The code structure should look similar to this:
Do the same using , use the `round()` function to round to 2.d.p:

```python
LIST = []
for INTEGER in range(X):
       value = CALCULATION
       value2 = ROUND value TO 2DP (using the above function)
       LIST.append(value2) #adds a new value to the end of the list
```

* Note that the append function adds to the end of a list as in this example (try it!)
```python
a = [1,2,"hello", pi]
a.append(42)
print(a)
```
`[1, 2, 'hello', 3.141592653589793, 42]`

In [31]:
##EDIT THE LINES BELOW (after removing a #) 

# LIST = []

#for INTEGER in range(X):
#    VALUE = CALCULATION
#    VALUE2 = round VALUE to 2dp (using the above function)
#    LIST.append(VALUE2) #adds a new value to the end of the list
i=1
LIST = (list(range(9)))

from math import pi
for INTEGER in range(9):
    i = (-pi + INTEGER*(2*pi)/8)
    w = "{:.2f}" .format(i)
    print(w, end = ", ")
    LIST.append(w)
    
print(LIST)


-3.14, -2.36, -1.57, -0.79, 0.00, 0.79, 1.57, 2.36, 3.14, [0, 1, 2, 3, 4, 5, 6, 7, 8, '-3.14', '-2.36', '-1.57', '-0.79', '0.00', '0.79', '1.57', '2.36', '3.14']


Result: `[-3.14, -2.36, -1.57, -0.79, 0.0, 0.79, 1.57, 2.36, 3.14]`  

HIDDEN SOLUTION  

<!--
mylist=[]

for n in range(9):  # n goes from 0 to 99
    x = -pi + n*2*pi/8  # formula for values
    x = round(x, 2)
    mylist.append(x)

print(mylist)
-->

Nested Loops
------------

We can also have loops inside loops, which are called <span>*nested
loops*</span>. In the case of a nested FOR loop, for each item in the
list we then iterate over the contents and perform some action.

In [10]:
for a_word in ["spam", "eggs", "cheese"]:
    print(a_word+":", end=' ')

    # this loop executes once per iteration of the outer loop
    for a_letter in a_word:
        print(a_letter, ";", end=' ')

    # "\n" starts a new line. could also use: print ""  without a comma (try this)
    print("\n", end=' ')

spam: s ; p ; a ; m ; 
 eggs: e ; g ; g ; s ; 
 cheese: c ; h ; e ; e ; s ; e ; 
 

Where the general format is:
```python
HEADER 1: 
    CODE FOR BLOCK 1
        HEADER 2: 
            CODE FOR BLOCK 2 
            ... 
        MORE CODE FOR BLOCK 1 
        ...
MAIN PROGRAMME
```

In [11]:
n = 3
for i in range(n):
    for j in range(1, 4):
        print(i, j, i*n + j)

0 1 1
0 2 2
0 3 3
1 1 4
1 2 5
1 3 6
2 1 7
2 2 8
2 3 9


Another useful function is `enumerate()`, which gives an index number to items in a list or string:

In [12]:
greeting = "Hello World!"
for idx, letr in enumerate(greeting):
    print(f"Character with index {idx} is: {letr}")

Character with index 0 is: H
Character with index 1 is: e
Character with index 2 is: l
Character with index 3 is: l
Character with index 4 is: o
Character with index 5 is:  
Character with index 6 is: W
Character with index 7 is: o
Character with index 8 is: r
Character with index 9 is: l
Character with index 10 is: d
Character with index 11 is: !


* Note that the `f"Some String with {variable}"` allows the inline formatting of a variable directly in a string.

# Task 3

## Creating a Simple Histogram

A histogram puts data into discrete compartments (*"bins"*) to look at the disribution of the number of data in each interval.

* Start a new notebook before doing this task.


1.  Create a data list to hold the following data points, which lie between 0 and 10:
    $$3.6, 4.5, 2.3, 2.6, 5.3, 4.9, 4.1, 8.1, 6.8, 6.7, 5.1, 5.7, 3.9, 5.4, 4.9, 5.1, 4.7, 4.7, 7.9, 3.8$$

2.  Use the method below to make a list of zeros of length 10 (to hold the count values):  
    `HISTLIST = [0]*10`  

3.  Loop through the data list, turning each value into an int, then
    add one to that element of the histogram array.(using the int as the
    index value):
    ```
    for each value in your data list:
         index_integer = #integer of the current value for which bin to add to
         HISTLIST[index_integer] += 1 #adds one to the specified value
    ``` 

4.  Print the histogram list.


* Download as a `.py` python file and check in Spyder or commandline before syubitting the `EXERCISE_2.py` to moodle. 

## Extra Exercises for more Practice

### Fibonacci Sequence Iteration

Use the following outline for a code (so-called *"pseudo-code"*) to write a program to generate the first
12 values of the Fibonacci sequence.

In [ ]:
# Fibonacci sequence

# 1. Define a new list containing first two values (1 and 1) using square bracket notation,

# 2. use a FOR loop and the range() function to iterate for 10 steps

# 3. use negative indexing to add the last two values of the list and
#   assign it to a variable for the next value

# 4. append this next value to the end of the list using syntax like:
# LISTNAME=LISTNAME+[NEWENTRY]

# 5. Print the list

Result: `[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144]`

-   Note that the new entry has to be enclosed in square brackets `[]`
    to convert it to a list. This is because lists can only be added to
    other lists.
    
-   Try printing the ratio of the last two steps at each iteration and
    see how they converge to the <span>*Golden Ratio*</span>.

[Hidden model solution, - double click here to see]

<!--#CODE:
# Fibonacci sequence
fib = [1, 1]  # first define the first two values

# loop for 10 steps
for i in range(10):
    nextval = fib[-1]+fib[-2]  # sum the last two values in the fib list
    fib = fib+[nextval]  # append the next value to the end of the list

print(fib)
-->

## Square Root Iteration 

The value of $\sqrt{2}$ can be calculated by iteration using: 
$x_n = \dfrac{1}{2}\left(x_{n-1} + \dfrac{2}{x_{n-1}}\right)$

Use the techniques you learned for the $\pi$ sequence to:

1.  Use a FOR loop to perform it for 10 steps

    -   Check the value against the actual value $x=\sqrt{2}=2^{0.5}$

In [ ]:
#initialise x to 2

#for loop goes here

print(x, 2**0.5)

Result: `1.414213562373095 1.4142135623730951`

Hidden solution:

<!--#code:
x = 2
for i in range(10):
    x = 0.5*(x+2/x)
print(x, 2**0.5)
-->

## Backwards Indexed Loop

Use the `range(START, STOP, STEP)` function with backwards indexing and a `for` loop to reverse the string `"Backwards!"`
    - Hint: you will need to experiment with the limits to get it right!

In [ ]:
b = "Backwards!"

#for loop:
    print(?, end="")

Result: `!sdrawkcaB`

Hidden solution:

<!--#Code:
b = "Backwards!"

for i in range(len(b)-1, -1, -1):
    print(b[i], end="")
-->